### DISCUSSION TITLE
Recurring critical SQLAlchemy errors are causing DB connection failure

### DISCUSSION THREAD

**Discussion by mmuht** [sqlalchemy/sqlalchemy#11727] [Usage Questions]
### Describe the bug

Hello Dears,

I am experiencing recurring critical SQLAlchemy errors as below, which are causing DB connection failure:

(sqlalchemy.dialects.postgresql.asyncpg.InterfaceError) \u003cclass 'asyncpg.exceptions._base.InterfaceError'\u003e: cannot call PreparedStatement.fetch(): the underlying connection is closed

(Background on this error at: https://sqlalche.me/e/14/rvf5)\n

Can't reconnect until invalid transaction is rolled back. (Background on this error at: https://sqlalche.me/e/14/8s2b)

I am using the below dependencies versions:

SQLAlchemy==1.4.39
asyncpg==0.26.0
fastapi==0.78.0
Python 3.9.0

Thank you for your support in advance!

### SQLAlchemy Version in Use

SQLAlchemy==1.4.39

### DBAPI (i.e. the database driver)

asyncpg==0.26.0

### Database Vendor and Major Version

PostgreSQL

### Python Version

Python 3.9.0

### Operating system

Linux

### To Reproduce

```python
import asyncio
import logging

from sqlalchemy.ext.asyncio import AsyncSession, create_async_engine
from sqlalchemy.orm import sessionmaker
from sqlalchemy.ext.declarative import declarative_base


logger = logging.getLogger(__name__)


Base = declarative_base()


def create_async_session() -> sessionmaker:
    db_engine = create_async_engine(DATABASE_URL, echo=False)
    return sessionmaker(db_engine, class_=AsyncSession, expire_on_commit=False)


async_session = create_async_session()


async def get_session() -> AsyncSession:

    try:
        async with async_session() as session:
            yield session

    except Exception as err:
        logger.error(f"Error occurred in get_session: {err}")
        raise err


def get_sync_db_session() -> AsyncSession:
    session = None

    try:
        async def get_db_session():
            async with async_session() as session:
                return session

        session = asyncio.get_event_loop().run_until_complete(get_db_session())
        return session

    except Exception as err:
        if session:
            session.rollback()

        logger.error(f"Error occurred in get_sync_db_session: {err}")
        raise err

    finally:
        if session:
            session.close()
```

### Error

```
(sqlalchemy.dialects.postgresql.asyncpg.InterfaceError) \u003cclass 'asyncpg.exceptions._base.InterfaceError'\u003e: cannot call PreparedStatement.fetch(): the underlying connection is closed

(Background on this error at: https://sqlalche.me/e/14/rvf5)\n

Can't reconnect until invalid transaction is rolled back. (Background on this error at: https://sqlalche.me/e/14/8s2b)
```

**Comment by CaselIT** [comment]
Hi,

It's not clear what you are trying to do. You cannot use an async session in a sync context, so I don't understand what's the purpose of `get_sync_db_session`

**Comment by mmuht** [comment]
@CaselIT Thank you for your feedback. I refactored the db session code as below:


import logging

from sqlalchemy.ext.asyncio import AsyncSession, create_async_engine
from sqlalchemy.orm import sessionmaker
from sqlalchemy.ext.declarative import declarative_base

Base = declarative_base()

def create_async_session() -> sessionmaker:
    db_engine = create_async_engine(DATABASE_URL, echo=False)
    return sessionmaker(db_engine, class_=AsyncSession, expire_on_commit=False)


async_session = create_async_session()


async def get_session() -> AsyncSession:

    try:
        async with async_session() as session:
            yield session

    except Exception as err:
        logger.error(f"Error occurred in get_session: {err}")
        raise err
        
        
However, I am still encountering the below errors. Any thoughts the reason behind this:
  
 (sqlalchemy.dialects.postgresql.asyncpg.InterfaceError) \u003cclass 'asyncpg.exceptions._base.InterfaceError'\u003e: cannot call PreparedStatement.fetch(): the underlying connection is closed

(Background on this error at: https://sqlalche.me/e/14/rvf5)\n

Can't reconnect until invalid transaction is rolled back. (Background on this error at: https://sqlalche.me/e/14/8s2b)
